File: alas/sidecar/schema.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Copyright (C) 2026 Marcos Quiroga Rodríguez

"""
Dataclass -> JSON-schema introspection.

Walks a config dataclass and emits a plain JSON description a React
``DynamicForm`` component can render generically, using field-metadata
conventions (``label``/``unit``/``help``/``min``/``max``/``decimals``/
``columns``/``advanced``/``hide_in_form``/``readonly_unless``) declared on
each dataclass field.

Deliberately independent of any GUI toolkit -- this module (and the rest of
``alas.sidecar``) must import cleanly headless.
"""

from __future__ import annotations

from dataclasses import fields, is_dataclass
from typing import Any, Dict, List, Tuple

from ..i18n import t

# Kept in sync with the frontend's own field-type table -- see DynamicForm if this
# one ever needs to change (e.g. a new unit suffix).
_UNIT_SUFFIXES: List[Tuple[str, str]] = [
    ("_kg_m2", "kg/m²"),
    ("_m_s2", "m/s²"),
    ("_m_s", "m/s"),
    ("_m2", "m²"),
    ("_deg", "deg"),
    ("_kg", "kg"),
    ("_pa", "Pa"),
    ("_m", "m"),
    ("_s", "s"),
]

_WEIGHT_SUFFIXES = ("_scale", "_per_m", "_weight", "_floor", "_cost", "_floor_m")


def _humanize(name: str) -> Tuple[str, str]:
    unit = ""
    label_src = name
    for suffix, u in _UNIT_SUFFIXES:
        if name.endswith(suffix):
            unit, label_src = u, name[: -len(suffix)]
            break
    label = label_src.replace("_", " ").strip().capitalize()
    return label, unit


def _is_number(v: Any) -> bool:
    return isinstance(v, (int, float)) and not isinstance(v, bool)


def _is_number_tuple(v: Any) -> bool:
    return isinstance(v, (list, tuple)) and len(v) > 0 and all(_is_number(x) for x in v)


def _string_field_extras(name: str, instance: Any) -> Dict[str, Any]:
    """Mirrors the frontend's string-field branch: which
    string fields get a dropdown, with what options, and whether free text
    is still allowed alongside the suggestions (Qt's ``setEditable(True)``).
    """
    if "airfoil" in name.lower():
        from ..geometry.airfoils import AirfoilLibrary

        return {"editable": True, "options": AirfoilLibrary.get_available_airfoils()}
    if name == "engine_name":
        # Same registry the Inputs screen's Engine dropdown is served from
        # (GET /engines) -- without this, the Advanced Settings geometry form
        # rendered engine_name as bare free text with no discoverability and
        # no protection against a typo that get_engine() then KeyErrors on.
        from ..config.engines import available_engines

        return {"editable": False, "options": available_engines()}
    # editable=False renders a strict <select> (the option list IS the valid
    # set -- free text could only produce a KeyError later); editable=True
    # renders a searchable combo box that still allows free text (airfoils:
    # AeroSandbox can also resolve names not in the library, e.g. any NACA
    # 4-digit code).
    if name.endswith("_material") and name != "strut_material":
        from ..config.materials import available_materials

        return {"editable": False, "options": available_materials()}
    if name == "strut_material":
        # Validated against physics.landing_gear.STRUT_MATERIALS (plus the
        # "auto" sentinel), not config.materials.MATERIAL_DATABASE -- its own
        # option list, not the generic `_material` one above.
        from ..physics.landing_gear import STRUT_MATERIALS

        options = ["auto"] + sorted(set(STRUT_MATERIALS.values()))
        return {"editable": False, "options": options}
    if name == "tire_class":
        from ..physics.landing_gear import TIRE_DATABASE, _TIRE_ORDER

        return {
            "editable": False,
            "options": ["auto"] + [k for k in _TIRE_ORDER if k in TIRE_DATABASE],
        }
    if name == "te_rib_mode":
        return {
            "editable": False,
            "options": ["all", "none", "alternate", "inboard", "outboard"],
        }
    if name == "strategy":
        # scipy.optimize.differential_evolution's built-in strategy names.
        return {
            "editable": False,
            "options": [
                "best1bin",
                "best1exp",
                "rand1bin",
                "rand1exp",
                "best2bin",
                "best2exp",
                "rand2bin",
                "rand2exp",
                "randtobest1bin",
                "randtobest1exp",
                "currenttobest1bin",
                "currenttobest1exp",
            ],
        }
    if name == "aircraft_type":
        return {"editable": False, "options": ["passenger", "cargo"]}
    if name == "cabin_preset":
        options_by_aircraft_type = {
            "passenger": ["Ryanair", "Iberia", "Emirates", "Custom"],
            "cargo": ["Max payload", "Dense payload", "Custom"],
        }
        aircraft_type = getattr(instance, "aircraft_type", "passenger")
        return {
            "editable": False,
            "options": options_by_aircraft_type.get(
                aircraft_type, options_by_aircraft_type["passenger"]
            ),
            "options_by_aircraft_type": options_by_aircraft_type,
        }
    return {}


def _field_kind(value: Any, name: str) -> str:
    if isinstance(value, bool):
        return "bool"
    if isinstance(value, int):
        return "int"
    if isinstance(value, float):
        return (
            "weight_slider"
            if any(name.endswith(s) for s in _WEIGHT_SUFFIXES) and value > 0
            else "float"
        )
    if isinstance(value, str):
        return "str"
    if value is None:
        return "optional"
    if isinstance(value, (list, tuple)) and value and all(_is_number(v) for v in value):
        return "number_list"
    if (
        isinstance(value, (list, tuple))
        and value
        and all(_is_number_tuple(v) for v in value)
    ):
        return "tuple_list"
    return "unsupported"


def dataclass_schema(instance: Any) -> Dict[str, Any]:
    """Return a JSON-safe schema dict for one (possibly nested) dataclass instance.

    Shape (one entry per field, in declaration order)::

        {
          "type": "<DataclassName>",
          "fields": [
            {"name", "kind", "label", "unit", "help", "advanced",
             "value", "min", "max", "decimals", "columns",
             ["fields": [...]] }   # present only when kind == "dataclass"
          ]
        }
    """
    if not is_dataclass(instance):
        raise TypeError("dataclass_schema requires a dataclass instance")

    out_fields: List[Dict[str, Any]] = []
    for f in fields(instance):
        meta = dict(f.metadata or {})
        if meta.get("hide_in_form"):
            continue
        value = getattr(instance, f.name)
        auto_label, auto_unit = _humanize(f.name)
        # Single translation boundary for every auto-generated form field: the
        # dataclasses keep their canonical English prose, and the active
        # language (set per request from ?lang=) is applied here on the way out.
        # An untranslated string falls through as English -- see alas/i18n.
        entry: Dict[str, Any] = {
            "name": f.name,
            "label": t(meta.get("label", auto_label)),
            "unit": meta.get("unit", auto_unit),
            "help": t(meta.get("help", "")),
            "advanced": bool(meta.get("advanced", False)),
        }
        if is_dataclass(value):
            entry["kind"] = "dataclass"
            entry["fields"] = dataclass_schema(value)["fields"]
        else:
            entry["kind"] = _field_kind(value, f.name)
            entry["value"] = value if entry["kind"] != "optional" else None
            if "min" in meta:
                entry["min"] = meta["min"]
            if "max" in meta:
                entry["max"] = meta["max"]
            if "decimals" in meta:
                entry["decimals"] = meta["decimals"]
            if "columns" in meta:
                entry["columns"] = meta["columns"]
            if "readonly_unless" in meta:
                entry["readonly_unless"] = meta["readonly_unless"]
            if entry["kind"] == "str":
                entry.update(_string_field_extras(f.name, instance))
        out_fields.append(entry)

    return {"type": type(instance).__name__, "fields": out_fields}

File: crates/alas-config/src/schema.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Marcos Quiroga Rodriguez

// Ported from alas/sidecar/schema.py
// Reference: alas @ rust-port-baseline.

//! The description a configuration struct gives of its own fields.
//!
//! This is what the settings interface is rendered from, and it is produced
//! by `#[derive(ConfigNode)]` rather than written by hand. The shape follows
//! the reference implementation's: an ordered list of fields, each carrying
//! what to call it, what unit it is in, what it means, and either its current
//! value or -- when the field is itself a configuration group -- the same
//! description one level down.
//!
//! Two things the reference does at this boundary are deliberately not done
//! here. It translates every label and help string on the way out, because
//! that function is its HTTP boundary and the language belongs to the
//! request; here the schema carries the canonical English and
//! [`Node::translated`] applies a language when something is about to show it
//! to a person, which is the arrangement `alas-i18n` already documents. It
//! also resolves the accepted values of the handful of string fields that
//! have them by importing the module that owns each list; those modules sit
//! above this crate in the layering, so a field names its list through
//! [`OptionSource`] and something that can see both resolves it.

use serde::ser::{SerializeMap, Serializer};
use serde::Serialize;

/// One configuration struct's fields, in declaration order.
#[derive(Debug, Clone, PartialEq)]
pub struct Node {
    /// The struct's name, as the interface titles the group.
    pub type_name: &'static str,
    /// Its fields, in the order they are declared.
    pub fields: Vec<Field>,
}

/// One field of a configuration struct.
#[derive(Debug, Clone, PartialEq)]
pub struct Field {
    /// The field's identifier, which is also its key in a saved file.
    pub name: &'static str,
    /// What to call it, in English.
    pub label: &'static str,
    /// Its unit, or the empty string when it is dimensionless.
    pub unit: &'static str,
    /// What it means, in English.
    pub help: &'static str,
    /// Whether the interface hides it until advanced settings are revealed.
    pub advanced: bool,
    /// Its value, or its own fields when it is a group.
    pub entry: Entry,
}

/// What sits under a field: a nested group, or a value.
#[derive(Debug, Clone, PartialEq)]
pub enum Entry {
    /// A nested configuration group.
    Node(Node),
    /// A value the interface edits directly.
    Leaf(LeafField),
}

/// A value-bearing field, with whatever constraints it declared.
#[derive(Debug, Clone, PartialEq)]
pub struct LeafField {
    /// Which editor the interface should use.
    pub kind: Kind,
    /// The current value.
    pub value: serde_json::Value,
    /// The smallest accepted value, if the field states one.
    pub min: Option<Number>,
    /// The largest accepted value, if the field states one.
    pub max: Option<Number>,
    /// Decimal places to show, if the field states a preference.
    pub decimals: Option<u32>,
    /// Column headings, for a field holding a table of numbers.
    pub columns: Option<&'static [&'static str]>,
    /// The sibling field and value that make this one editable.
    pub readonly_unless: Option<ReadonlyUnless>,
    /// Where this field's accepted values come from.
    pub options: Option<OptionSource>,
}

/// A bound, which keeps whether it was written as an integer.
///
/// A seat count's minimum is `0` and a seat pitch's is `0.7112`; serializing
/// the first as `0.0` would make a whole-number field look like a fractional
/// one to whatever renders it.
#[derive(Debug, Clone, Copy, PartialEq, Serialize)]
#[serde(untagged)]
pub enum Number {
    /// A whole-number bound.
    Integer(i64),
    /// A fractional bound.
    Real(f64),
}

impl From<i64> for Number {
    fn from(value: i64) -> Self {
        Self::Integer(value)
    }
}

impl From<f64> for Number {
    fn from(value: f64) -> Self {
        Self::Real(value)
    }
}

/// A field that is editable only while a sibling holds a particular value.
///
/// The sibling is named rather than referenced: the condition is resolved
/// against the values the form currently holds, which upstream resolves at
/// the nearest enclosing level rather than necessarily in the same struct.
#[derive(Debug, Clone, Copy, PartialEq, Serialize)]
pub struct ReadonlyUnless {
    /// The sibling field's name.
    pub field: &'static str,
    /// The value it must hold.
    pub value: &'static str,
}

/// Which editor a value gets.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum Kind {
    /// A checkbox.
    Bool,
    /// A whole number.
    Int,
    /// A real number.
    Float,
    /// A real number the interface offers as a slider, because it is a
    /// relative weight rather than a physical quantity.
    WeightSlider,
    /// Text, possibly with a list of accepted values.
    Str,
    /// A value that is currently unset.
    Optional,
    /// A list of numbers.
    NumberList,
    /// A list of number pairs -- a small table.
    TupleList,
    /// A nested configuration group.
    ///
    /// Spelled as the reference spells it, because the parity fixture is its
    /// output and a renamed variant would mean the comparison is against a
    /// translation of that output rather than against it.
    #[serde(rename = "dataclass")]
    Nested,
    /// A value with no editor. Reproduces the reference's classification,
    /// which reaches this for an empty list because it tests the list's
    /// contents and an empty one has none.
    Unsupported,
}

/// Where a string field's accepted values come from.
///
/// Most of these lists are owned by crates that sit above this one -- the
/// airfoil library, the engine deck, the material database -- so this names
/// the list and something that can see both resolves it. The lists that
/// depend on nothing are resolved by [`OptionSource::options`] here.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum OptionSource {
    /// The airfoil library, which also accepts names it does not list: any
    /// NACA 4-digit code resolves without being in it.
    Airfoil,
    /// The engine registry.
    Engine,
    /// The material database.
    Material,
    /// The landing-gear strut materials, which are their own list and not
    /// the general material database.
    StrutMaterial,
    /// The tire database.
    TireClass,
    /// Which trailing-edge ribs the structural mesh generates.
    TeRibMode,
    /// The differential-evolution strategy names.
    Strategy,
    /// The top-level aircraft optimization algorithms.
    OptimizerMethod,
    /// Whether the aircraft carries passengers or freight.
    AircraftType,
    /// The cabin layout presets, which differ by aircraft type.
    CabinPreset,
    /// The one method that owns every production mass group.
    MassArchitecture,
    /// What kind of knowledge a family of declared FLOPS inputs rests on.
    FlopsInputEvidence,
    /// Versioned systems-and-equipment mass method.
    SystemsMassMethod,
    /// Versioned structural-group mass method.
    StructuralMassMethod,
    /// Versioned propulsion-group mass method.
    PropulsionMassMethod,
    /// Which FLOPS wing bending-material factor is evaluated.
    FlopsWingBendingMethod,
    /// The operating rule a design mission's reserves are sized under.
    FuelScheme,
    /// The scalar the mission-sized design search minimises.
    ObjectiveKind,
    /// Whether the takeoff mass is a fixed input, closed by the mission up
    /// to it, or closed by the mission with it used only to seed the first
    /// pass.
    MtowSizing,
    /// How a family of requirements takes part in the ranking.
    ConstraintPolicy,
    /// How the optimizer treats the aircraft geometry it starts from.
    DesignMode,
}

impl OptionSource {
    /// The accepted values, when this crate can name them.
    ///
    /// `None` means the list is owned elsewhere and has to be resolved by a
    /// crate that can see its owner.
    pub fn options(self) -> Option<&'static [&'static str]> {
        match self {
            Self::TeRibMode => Some(&["all", "none", "alternate", "inboard", "outboard"]),
            Self::Strategy => Some(&[
                "best1bin",
                "best1exp",
                "rand1bin",
                "rand1exp",
                "best2bin",
                "best2exp",
                "rand2bin",
                "rand2exp",
                "randtobest1bin",
                "randtobest1exp",
                "currenttobest1bin",
                "currenttobest1exp",
            ]),
            Self::OptimizerMethod => Some(&[
                "differential_evolution",
                "feasibility_first_de",
                "nsga2",
                "turbo_1",
                "cma_es",
                "sqp",
            ]),
            Self::AircraftType => Some(&["passenger", "cargo"]),
            Self::MassArchitecture => Some(&[
                "pure_flops_transport_v1",
                "legacy_reference_compatible_comparison",
            ]),
            Self::FlopsInputEvidence => Some(&[
                "source_backed",
                "user_declared",
                "published_flops_default",
                "uncertain_engineering_estimate",
            ]),
            Self::SystemsMassMethod => {
                Some(&["reference_compatible_fractions", "flops_transport_v1"])
            }
            Self::StructuralMassMethod | Self::PropulsionMassMethod => {
                Some(&["reference_compatible", "flops_transport_v1"])
            }
            Self::FlopsWingBendingMethod => Some(&["simplified", "detailed"]),
            Self::FuelScheme => Some(&[
                "easa_basic",
                "faa_domestic",
                "faa_flag_supplemental",
                "study_convention",
                "trip_fuel_only",
            ]),
            Self::ObjectiveKind => Some(&[
                "block_fuel",
                "takeoff_mass",
                "operating_empty_mass",
                "fuel_per_seat_kilometre",
            ]),
            Self::MtowSizing => Some(&["fixed_requirement", "sized_by_mission", "unconstrained"]),
            Self::ConstraintPolicy => Some(&["hard", "soft", "diagnostic", "off"]),
            Self::DesignMode => Some(&["clean_sheet", "reference_adaptation", "baseline_sandbox"]),
            _ => None,
        }
    }

    /// Whether a value outside the list is still accepted.
    ///
    /// Only the airfoil field is: the geometry layer resolves names the
    /// library does not carry, so a strict list would reject valid input.
    /// Everywhere else the list is the valid set, and free text could only
    /// produce a lookup failure later.
    pub fn editable(self) -> bool {
        self == Self::Airfoil
    }
}

impl Node {
    /// This node with every label and help string looked up in `lang`.
    ///
    /// Untranslated strings stay as they are, so a partial catalog degrades
    /// to English rather than to a missing-key placeholder. Returns owned
    /// strings, since a translation is not a `&'static str`.
    pub fn translated(&self, lang: Option<&str>) -> TranslatedNode {
        TranslatedNode {
            type_name: self.type_name,
            fields: self
                .fields
                .iter()
                .map(|field| TranslatedField {
                    name: field.name,
                    label: alas_i18n::t(Some(field.label), lang).into_owned(),
                    unit: field.unit,
                    help: alas_i18n::t(Some(field.help), lang).into_owned(),
                    advanced: field.advanced,
                    entry: match &field.entry {
                        Entry::Node(node) => TranslatedEntry::Node(node.translated(lang)),
                        Entry::Leaf(leaf) => TranslatedEntry::Leaf(leaf.clone()),
                    },
                })
                .collect(),
        }
    }

    /// The field of this node with the given name, if it has one.
    pub fn field(&self, name: &str) -> Option<&Field> {
        self.fields.iter().find(|field| field.name == name)
    }
}

/// A [`Node`] whose prose has been looked up in a language.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct TranslatedNode {
    /// The struct's name.
    pub type_name: &'static str,
    /// Its fields, in declaration order.
    pub fields: Vec<TranslatedField>,
}

/// A [`Field`] whose prose has been looked up in a language.
#[derive(Debug, Clone, PartialEq)]
pub struct TranslatedField {
    /// The field's identifier.
    pub name: &'static str,
    /// What to call it, in the active language.
    pub label: String,
    /// Its unit.
    pub unit: &'static str,
    /// What it means, in the active language.
    pub help: String,
    /// Whether it is hidden until advanced settings are revealed.
    pub advanced: bool,
    /// Its value, or its own fields.
    pub entry: TranslatedEntry,
}

/// What sits under a [`TranslatedField`].
#[derive(Debug, Clone, PartialEq)]
pub enum TranslatedEntry {
    /// A nested group.
    Node(TranslatedNode),
    /// A value.
    Leaf(LeafField),
}

impl Serialize for Node {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut map = serializer.serialize_map(Some(2))?;
        map.serialize_entry("type", self.type_name)?;
        map.serialize_entry("fields", &self.fields)?;
        map.end()
    }
}

impl Serialize for Field {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut map = serializer.serialize_map(None)?;
        map.serialize_entry("name", self.name)?;
        map.serialize_entry("label", self.label)?;
        map.serialize_entry("unit", self.unit)?;
        map.serialize_entry("help", self.help)?;
        map.serialize_entry("advanced", &self.advanced)?;
        match &self.entry {
            Entry::Node(node) => {
                map.serialize_entry("kind", &Kind::Nested)?;
                map.serialize_entry("fields", &node.fields)?;
            }
            Entry::Leaf(leaf) => leaf.serialize_into(&mut map)?,
        }
        map.end()
    }
}

impl Serialize for TranslatedField {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let mut map = serializer.serialize_map(None)?;
        map.serialize_entry("name", self.name)?;
        map.serialize_entry("label", &self.label)?;
        map.serialize_entry("unit", self.unit)?;
        map.serialize_entry("help", &self.help)?;
        map.serialize_entry("advanced", &self.advanced)?;
        match &self.entry {
            TranslatedEntry::Node(node) => {
                map.serialize_entry("kind", &Kind::Nested)?;
                map.serialize_entry("fields", &node.fields)?;
            }
            TranslatedEntry::Leaf(leaf) => leaf.serialize_into(&mut map)?,
        }
        map.end()
    }
}

impl LeafField {
    /// Write this value's keys into an already-open field map.
    ///
    /// A constraint the field did not declare is an absent key rather than a
    /// null, which is how the reference emits it: the interface tells "no
    /// minimum" from "a minimum of nothing" by the key's presence.
    fn serialize_into<M: SerializeMap>(&self, map: &mut M) -> Result<(), M::Error> {
        map.serialize_entry("kind", &self.kind)?;
        map.serialize_entry("value", &self.value)?;
        if let Some(min) = &self.min {
            map.serialize_entry("min", min)?;
        }
        if let Some(max) = &self.max {
            map.serialize_entry("max", max)?;
        }
        if let Some(decimals) = &self.decimals {
            map.serialize_entry("decimals", decimals)?;
        }
        if let Some(columns) = &self.columns {
            map.serialize_entry("columns", columns)?;
        }
        if let Some(readonly_unless) = &self.readonly_unless {
            map.serialize_entry("readonly_unless", readonly_unless)?;
        }
        if let Some(options) = &self.options {
            map.serialize_entry("options_from", options)?;
            map.serialize_entry("editable", &options.editable())?;
            if let Some(values) = options.options() {
                map.serialize_entry("options", values)?;
            }
        }
        Ok(())
    }
}
